src/ray/gcs/gcs_server/test/gcs_placement_group_manager_mock_test.cc

```diff
@@ -0,0 +1,135 @@
+      .WillOnce(DoAll(SaveArg<3>(&put_cb), Return(Status::OK())));
+  EXPECT_CALL(*gcs_placement_group_scheduler_, ScheduleUnplacedBundles(_, _, _))
+      .WillOnce(DoAll(SaveArg<1>(&failure_callback), SaveArg<2>(&success_callback)));
+  auto now = absl::GetCurrentTimeNanos();
+  gcs_placement_group_manager_->RegisterPlacementGroup(pg, cb);
+  auto &pending_queue = gcs_placement_group_manager_->pending_placement_groups_;
+  ASSERT_EQ(1, pending_queue.size());
+  ASSERT_LE(now, pending_queue.begin()->first);
+  ASSERT_GE(absl::GetCurrentTimeNanos(), pending_queue.begin()->first);
+  put_cb(Status::OK());
+  pg->UpdateState(rpc::PlacementGroupTableData::RESCHEDULING);
+  failure_callback(pg, true);
+  ASSERT_EQ(1, pending_queue.size());
+  ASSERT_GE(0, pending_queue.begin()->first);
+}
+
+TEST_F(GcsPlacementGroupManagerMockTest, PendingQueuePriorityFailed) {
+  // Test priority works
+  //   When return with a failure, exp backoff should work
+  auto req =
+      Mocker::GenCreatePlacementGroupRequest("", rpc::PlacementStrategy::SPREAD, 1);
+  auto pg = std::make_shared<GcsPlacementGroup>(req, "");
+  auto cb = [](Status s) {};
+  PGSchedulingFailureCallback failure_callback;
+  PGSchedulingSuccessfulCallback success_callback;
+  StatusCallback put_cb;
+  EXPECT_CALL(*store_client_, AsyncPut(_, _, _, _))
+      .WillOnce(DoAll(SaveArg<3>(&put_cb), Return(Status::OK())));
+  EXPECT_CALL(*gcs_placement_group_scheduler_, ScheduleUnplacedBundles(_, _, _))
+      .Times(2)
+      .WillRepeatedly(
+          DoAll(SaveArg<1>(&failure_callback), SaveArg<2>(&success_callback)));
+  auto now = absl::GetCurrentTimeNanos();
+  gcs_placement_group_manager_->RegisterPlacementGroup(pg, cb);
+  auto &pending_queue = gcs_placement_group_manager_->pending_placement_groups_;
+  ASSERT_EQ(1, pending_queue.size());
+  ASSERT_LE(now, pending_queue.begin()->first);
+  ASSERT_GE(absl::GetCurrentTimeNanos(), pending_queue.begin()->first);
+  put_cb(Status::OK());
+  pg->UpdateState(rpc::PlacementGroupTableData::PENDING);
+  now = absl::GetCurrentTimeNanos();
+  failure_callback(pg, true);
+  auto exp_backer = ExponentialBackOff(
+      1000000 * RayConfig::instance().gcs_create_placement_group_retry_min_interval_ms(),
+      RayConfig::instance().gcs_create_placement_group_retry_multiplier(),
+      1000000 * RayConfig::instance().gcs_create_placement_group_retry_max_interval_ms());
+  auto next = exp_backer.Next();
+  ASSERT_DOUBLE_EQ(
+      next,
+      1000000 * RayConfig::instance().gcs_create_placement_group_retry_min_interval_ms());
+  ASSERT_EQ(1, pending_queue.size());
+  auto rank = pending_queue.begin()->first;
+  ASSERT_LT(now + next, rank);
+  // ScheduleUnplacedBundles is not called here
+  gcs_placement_group_manager_->SchedulePendingPlacementGroups();
+  ASSERT_EQ(1, pending_queue.size());
+  ASSERT_EQ(rank, pending_queue.begin()->first);
+
+  absl::SleepFor(absl::Nanoseconds(rank - absl::GetCurrentTimeNanos()));
```

Comment:
I actually feel this is fine. Using the fake timer in the code just for testing makes the code complicated.
It's less than 1s here.

Reply:
Can you at least use wait for ...? Based on my experience having a absolute time sleep usually was the source of flakniess

Reply:
If 
```
  absl::SleepFor(absl::Nanoseconds(rank - absl::GetCurrentTimeNanos()));
```
will guarentee  `rank - absl::GetCurrentTimeNanos() + e (e >= 0)` suspension of the code, it's ok. `e` is the scheduling latency here.
I don't think wait for is a better choice here. 
How about this, let me run it for 100 times and if it's good, we are good here?

Reply:
Let me justify what I mean here and let me know if it doesn't make sense and I can change it to use wait for.


Here if we `sleep s` seconds, usually it'll be `s + e`, where `e` equals the delay of os scheduling. 

So after executing this line, the time must be greater than rank. 

So in 125, it'll start to schedule the first one.

in L129, we'll have new rank  `r` where `r = t(the time when we push it back to queue) + (next) failed_delay`. `t > now` so `t + next = r > now + next` (L132).


Here the trick I'm using is that I use LT to compare the time, so if the theory stands, it's ok.

Reply:
Gotcha. I think it makes sense, and it is less likely that there's unexpected flakiness because it has pretty straightforward dependencies (unlike Ray python tests which has so many unexpected scenarios). I am okay with keeping it